doc/meson.build

```diff
@@ -0,0 +1,128 @@
+                      command: cmd)
+    endif
+    # geany.html is installed as index.html
+    custom_target('index.html',
+                  input: [geany_html],
+                  output: ['index.html'],
+                  command: [ln, '-f', '@INPUT@', '@OUTPUT@'],
+                  install: true,
+                  install_dir: join_paths(cdata.get('GEANY_DOC_DIR'), 'html'))
+endif
+
+install_data('geany.txt',
+             rename: 'manual.txt',
+             install_dir: cdata.get('GEANY_DOC_DIR'))
+install_data('../scintilla/License.txt',
+             rename: 'ScintillaLicense.txt',
+             install_dir: cdata.get('GEANY_DATA_DIR'))
+install_data('../scintilla/lexilla/License.txt',
+             rename: 'LexillaLicense.txt',
+             install_dir: cdata.get('GEANY_DATA_DIR'))
+
+# Normally, doxygen is required for the gtkdoc headers
+# but it can be disabled if really neded (e.g. if plugins are also disabled),
+# packagers must not disable this!
+doxygen = find_program('doxygen', required: false)
+if doxygen.found()
+    doxcfg = configuration_data()
+    doxcfg.merge_from(cdata)
+    doxcfg.set('GIRONLY', '@internal')
+    doxcfg.set('HTML', 'YES')
+    doxcfg.set('XML', 'NO')
+    doxcfg.set('SORT', 'YES')
+
+    doxgicfg = doxcfg
+    doxgicfg.set('GIRONLY', '')
+    doxgicfg.set('HTML', 'NO')
+    doxgicfg.set('XML', 'YES')
+    doxgicfg.set('SORT', 'NO')
+
+    dox = configure_file(input: 'Doxyfile.in',
+                         output: 'Doxyfile',
+                         configuration: doxcfg)
+    doxgi = configure_file(input: 'Doxyfile.in',
+                           output: 'Doxyfile-gi',
+                           configuration: doxgicfg)
+
+    dep_doxygen = declare_dependency(sources: ['plugins.dox',
+                                               'pluginsignals.c',
+                                               'pluginsymbols.c',
+                                               'stash-example.c',
+                                               'stash-gui-example.c'])
+
+    custom_target('doxygen.stamp',
+                  input: dox,
+                  output: ['doxygen.stamp'],
+                  command: [doxygen, '@INPUT@', '&&', 'touch', '@OUTPUT@'],
+                  depends: libgeany,
+                  build_by_default: true)
+    doxy_xml = custom_target('doxygen-gi.stamp',
+                             input: doxgi,
+                             output: ['doxygen-gi.stamp'],
+                             command: [doxygen, '@INPUT@', '&&', 'touch', '@OUTPUT@'],
+                             depends: libgeany,
+                             build_by_default: true)
+    custom_target('gtkdoc headers',
+                  input : [doxy_xml],
+                  output: ['geany-gtkdoc.h', 'geany-sciwrappers-gtkdoc.h'],
+                  command: [python, '../scripts/gen-api-gtkdoc.py', join_paths(meson.current_build_dir(), 'xml'), '-d', '.', '-o', '@OUTPUT0@', '--sci-output', '@OUTPUT1@'],
```

Comment:
I thought `../scripts/gen-api-gtkdoc.py` would be relative to the `doc/` folder where this meson.build is located. But for commands it seems to be relative to the build directory.

I also do 'install_data("../scintilla/…")`  further down in the same meson.build and that seems to work.

Reply:
When you run `cd builddir/ && ninja`, ninja is being run from the build directory.

And then a ninja rule contains the literal string `../scripts/gen-api-gtkdoc.py`, as you wrote.

find_program() will indeed make it relative to the doc/ folder where this meson.build is located, because find_program converts strings (passed as argv1 to the python executable) into found programs. Thus, the current working directory of the ***python*** program is the build directory.

`install_data()` accepts a str object and interprets it as a file path, hence install_data() converts strings into found resources and makes them relative to meson.build.

So does any other meson function where the context of the function specifies that the string argument should be interpreted as files/directories.

Again, this is not the case for `command`, which specifies that the first element `python` of the command is a file (program) and the remaining elements of the command such as `'../scripts/gen-api-gtkdoc.py'` are unknown and unknowable strings passed without modification.

Those unknown and unknowable strings, `'../'` included, are then interpreted by the python program, which happens to assume they represent a filepath.